### PR TITLE
Add LoRA support for FT API

### PR DESCRIPTION
**Pull request by artek0chumak**
Issue ENG-4853

Add LoRA training support for finetuning jobs.

Some cli examples:

```bash
together fine-tuning create --training-file "${FILE_ID}" --model "meta-llama/Meta-Llama-3-8B" --wandb-api-key "${WANDB_API_KEY}" --lora --lora-r 8
```

Description of the new parameters:

- `--lora`(bool) -- general flag for enabling LoRA training
- `--lora-r`(int) -- rank for LoRA adapter weights
- `--lora-dropout`(float) -- dropout value for LoRA adapter training
- `--lora-alpha`(int) -- alpha value for LoRA adapter training
- `--lora-trainable-modules`(str) -- LoRA adapters' trainable modules, separated by a comma. To use trainable modules, use `all-linear`

### CODE REVIEW

**Review by orangetin — APPROVED** [review]
LGTM

File: src/together/cli/api/finetune.py [hunk 1]
```diff
@@ -26,7 +27,17 @@ def fine_tuning(ctx: click.Context) -> None:
     "--n-checkpoints", type=int, default=1, help="Number of checkpoints to save"
 )
 @click.option("--batch-size", type=int, default=32, help="Train batch size")
-@click.option("--learning-rate", type=float, default=3e-5, help="Learning rate")
+@click.option("--learning-rate", type=float, default=1e-5, help="Learning rate")
+@click.option("--lora/--no-lora", type=bool, default=False, help="LoRA mode")
```
Comment (mryab) [inline src/together/cli/api/finetune.py:31]:
```suggestion
@click.option("--lora/--no-lora", type=bool, default=False, help="Whether to use LoRA adapters for fine-tuning")
```

File: src/together/cli/api/finetune.py [hunk 2]
```diff
@@ -26,7 +27,17 @@ def fine_tuning(ctx: click.Context) -> None:
...
+@click.option("--lora-r", type=int, default=8, help="LoRA adapters' rank")
+@click.option("--lora-dropout", type=float, default=0, help="LoRA adapters' dropout")
+@click.option("--lora-alpha", type=float, default=8, help="LoRA adapters' alpha")
+@click.option(
+    "--lora-trainable-modules",
+    type=str,
+    default="all-linear",
+    help="LoRA adapters' trainable modules. For example, 'all-linear', 'q_proj,v_proj'",
```
Comment (mryab) [inline src/together/cli/api/finetune.py:39]:
```suggestion
    help="Trainable modules for LoRA adapters. For example, 'all-linear', 'q_proj,v_proj'",
```

File: src/together/resources/finetune.py [hunk 3]
```diff
@@ -45,6 +52,11 @@ def create(
             batch_size (int, optional): Batch size for fine-tuning. Defaults to 32.
             learning_rate (float, optional): Learning rate multiplier to use for training
                 Defaults to 0.00001.
+            lora (bool, optional): Whether to use LoRA adapters. Defaults to True.
+            lora_r (int, optional): Whether to use LoRA R. Defaults to 8.
```
Comment (mryab) [inline src/together/resources/finetune.py:56]:
```suggestion
            lora_r (int, optional): Rank of LoRA adapters. Defaults to 8.
```

File: src/together/cli/api/finetune.py (hunk 1 above)
Comment (mryab) [inline src/together/cli/api/finetune.py:31]:
```suggestion
@click.option("--lora/--no-lora", type=bool, default=False, help="Whether to use LoRA adapters")
```

File: src/together/types/finetune.py [hunk 4]
```diff
@@ -121,6 +146,7 @@ class FinetuneRequest(BaseModel):
     suffix: str | None = None
     # weights & biases api key
     wandb_key: str | None = None
+    training_type: FullTrainingType | LoRATrainingType | None = None
```
Comment (mryab) [inline src/together/types/finetune.py:149]: Best to change it to just `TrainingType` perhaps? Or something like https://docs.python.org/3/library/typing.html#newtype to ensure that only subclasses of TrainingType can be used here

Comment (artek0chumak) [reply src/together/types/finetune.py:149]: I'm not sure `pydantic` will serialize NewType or TrainingType to right typings
